contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,458 @@
+
+#### Mount and Unmount
+
+The in-tree volume plugin’s SetUp and TearDown methods will trigger the `NodePublishVolume` and `NodeUnpublishVolume` CSI calls via Unix Domain Socket. Kubernetes will generate a unique `target_path` (unique per pod per volume) to pass via `NodePublishVolume` for the CSI plugin to mount the volume. Upon successful completion of the `NodeUnpublishVolume` call (once volume unmount has been verified), Kubernetes will delete the directory.
+
+The Kubernetes volume sub-system does not currently support block volumes (only file), so for alpha, the Kubernetes CSI volume plugin will only support file.
+
+#### Attaching and Detaching
+
+The attach/detach controller,running as part of the kube-controller-manager binary on the master, decides when a CSI volume must be attached or detached from a particular node.
+
+When the controller decides to attach a CSI volume, it will call the in-tree CSI volume plugin’s attach method. The in-tree CSI volume plugin’s attach method will do the following:
+
+1. Create a new `VolumeAttachment` object (defined in the “Communication Channels” section) to attach the volume.
+    * The name of the of the `VolumeAttachment` object will be the `pv-<PVName>-<NodeName>`.
+      * `pv-` prefix is used to allow using other scheme(s) for inline volumes in future, with their own prefix.
+      * `PVName` is `PV.name` of the attached PersistentVolume.
+      * `NodeName` is `Node.name` of the node where the volume should be attached to.
+    * If a `VolumeAttachment` object with the corresponding name already exists, the in-tree volume plugin will simply begin to poll it as defined below. The object is not modified; only the external-attacher should change the status fields; and the external-attacher is responsible for it’s own retry and error handling logic.
+2. Poll the `VolumeAttachment` object waiting for one of the following conditions:
+    * The `VolumeAttachment.Status.Attached` field to become `true`.
+      * The operation completes successfully.
+    * An error to be set in the `VolumeAttachment.Status.AttachError` field.
+      * The operation terminates with the specified error.
+    * The operation to timeout.
+      * The operation terminates with timeout error.
+    * The `VolumeAttachment.DeletionTimestamp` is set.
+      * The operation terminates with an error indicating a detach operation is in progress.
+      * The `VolumeAttachment.Status.Attached` value must not be trusted. The attach/detach controller has to wait until the object is deleted by the external-attacher before creating a new instance of the object.
+
+When the controller decides to detach a CSI volume, it will call the in-tree CSI volume plugin’s detach method. The in-tree CSI volume plugin’s detach method will do the following:
+
+1. Delete the corresponding `VolumeAttachment` object (defined in the “Communication Channels” section) to indicate the volume should be detached.
```

Comment:
Is the `VolumeAttachment` object to be `deleted` or have it's status updated ?  If it is `deleted` how will external attacher be notified (since object is deleted, reconciliation loop)?  Also, If object is actually deleted, object will not show up in future API server query (this may cause erroneous interpretation of a missing volumeattachment object).

Reply:
before external attacher does anything with the object, it adds a finalizer.

when the object is attempted to be deleted, if there are pending finalizers, it gets a deletionTimestamp but persists until all finalizers are removed.